std_traffic/utils/aws_utils.py

```diff
@@ -0,0 +1,75 @@
+"""Utility functions to use AWS services"""
+
+import logging
+from os import environ
+
+import boto3
+from botocore.exceptions import ClientError
+from boto3.s3.transfer import TransferConfig
+
+
+def _init_client(client_name: str) -> boto3.client:
+    return boto3.client(
+        client_name,
+        aws_access_key_id=environ['AWS_KEY_ID'],
+        aws_secret_access_key=environ['AWS_SECRET_KEY'],
+        region_name=environ['AWS_REGION'])
+
+
+def get_s3_client() -> boto3.client:
+    """Create a connection client to S3."""
+
+    return _init_client('s3')
+
```

Comment:
Why this and not use directly boto3.client? Security?

Reply:
@RuggeroFabbiano There are multiple clients exposed by the `boto3` package, and the only thing that changes to initialize them is the name. So, with the code above you could do

```
_init_client('s3')
```

but also

```
_init_client('sns')
```

Using `boto3.client` would require to always give the credentials (or to have them in the "famous" `/home/.awsconfig` file, but I don't like that solution).

What alternative do you recommend?

Reply:
I was meaning, in `_upload_file` you call `client = get_s3_client()`, that calls `_init_client('s3')`, that calls `boto3.client(args)`;
why not call in `client = boto3.client(args)` directly?

Reply:
The idea of `get_s3_client` is that it doesn't need any input and therefore has more control over what you get from it.
In another situation, I would tell you that the module hides the implementation of the `_init_client` because the user is not supposed to change that. Of course, in Python this is trash talking.
I still think it's nice anyway not to have to pass all the `args` around.
Do you have a suggestion to improve this part?

Reply:
Yes I guess I see the point (some sort of "fake encapsulation" if I get it right), but since `get_s3_client()` is only used in an already "private method" it seems to me that the implementation is already hidden in some sense, and all these levels can be avoided. It would be different if it was "public". Moreover, this would avoid creating a "public" `get_s3_client()` that can be used by the user (thus making the need to protect `_init_client()`). Is that wrong?

Reply:
I think you are suggesting to remove `_init_client(...)` and have instead that code inside `get_s3_client()`, like this:

```
def get_s3_client(**kwargs) -> boto3.client:
    aws_access_key_id = kwargs.get('AWS_KEY_ID', environ['AWS_KEY_ID'])
    aws_secret_access_key = kwargs.get('AWS_SECRET_KEY', environ['AWS_SECRET_KEY'])
    region_name = kwargs.get('AWS_REGION', environ['AWS_REGION'])
    return boto3.client(
        's3',
        aws_access_key_id=aws_access_key_id,
        aws_secret_access_key=aws_secret_access_key,
        region_name=region_name)
```

I don't like this because then, one day, I will have to write a `get_sns_client(...)` function, for a different boto3 client service, and I will have to rewrite the same 8 lines, just changing the first argument of the last call to `'sns'`.
Whereas, with the intermediate function `_init_client` I could just do

```
def get_sns_client(**kwargs):
    return _init_client('sns', **kwargs)
```

Maybe the trailing underscore of `_init_client` is confusing: my point here is more about avoiding duplicate code, much less about function visibility.

Reply:
I would more keep just `_init_client()` and use it directly into `_update_file()`, so one day you can just call `_init_client()` with `'sns'`.

So just to summarise, `_init_client()` helps to DRY code, whereas get_x_client() is used to "hide" what you're doing underneath, right?